src/librbd/api/Mirror.cc

```diff
@@ -756,6 +757,188 @@ void Mirror<I>::image_demote(I *ictx, Context *on_finish) {
+    lderr(cct) << "remote pool does not exist" << cpp_strerror(r) << dendl;
+    return r;
+  } else if (r < 0) {
+    lderr(cct) << "failed to open remote pool '" << io_ctx.get_pool_name()
+               << "': " << cpp_strerror(r) << dendl;
+    return r;
+  }
+  std::vector<mirror_peer_site_t> remote_peers;
+  r = Mirror<>::peer_site_list(remote_io_ctx, &remote_peers);
+  if (r < 0 && r != -ENOENT) {
+    lderr(cct) << "failed to list mirror remote peers: "
+               << cpp_strerror(r) << dendl;
+    return r;
+  }
+  if (remote_peers.empty()) {
+    lderr(cct) << "no peers found on remote" << dendl;
+    return -EINVAL;
+  }
+  ceph_assert(remote_peers.size() == 1);
+
+  auto remote_peer = *remote_peers.begin();
+  ldout(cct, 20) << "remote's peer deatils: "
+                 << ", UUID: " << remote_peer.uuid
+                 << ", site name: " << remote_peer.site_name
+                 << ", mirror UUID: " << remote_peer.mirror_uuid
+                 << ", client: " << remote_peer.client_name
+                 << dendl;
+
+  uint64_t size;
+  {
+    std::shared_lock image_locker{ictx->image_lock};
+    size =  ictx->get_image_size(CEPH_NOSNAP);
+  }
+  ldout(cct, 20) << "local image size: " << size << dendl;
+  uint64_t object_count = Striper::get_num_objects(ictx->layout, size);
+  ldout(cct, 20) << "local image object count: " << object_count << dendl;
+
+  librbd::ImageCtx *remote_ictx = new librbd::ImageCtx(ictx->name.c_str(), "",
+                                                       nullptr,
+                                                       remote_io_ctx, false);
+  remote_ictx->read_only_mask &= ~IMAGE_READ_ONLY_FLAG_NON_PRIMARY;
+  r = remote_ictx->state->open(0);
+  if (r < 0) {
+    lderr(cct) << "error opening image: " << ictx->name.c_str() << ": "
+               << cpp_strerror(r) << dendl;
+    return r;
+  }
+  uint64_t remote_size;
+  {
+    std::shared_lock image_locker{remote_ictx->image_lock};
+    remote_size =  remote_ictx->get_image_size(CEPH_NOSNAP);
+  }
+  ldout(cct, 20) << "remote image size: " << remote_size  << dendl;
+  uint64_t remote_object_count = Striper::get_num_objects(remote_ictx->layout,
+                                                          remote_size);
+  ldout(cct, 20) << "remote image object count: "
+                 << remote_object_count << dendl;
+
+  bufferlist init_value_bl;
+  bufferlist csum_bl;
+  bufferlist rcsum_bl;
+  bool local_obj_absent;
+  bool remote_obj_absent;
+
+  ceph::encode(static_cast<uint32_t>(-1), init_value_bl);
+  for (uint64_t i = 0; i < object_count; ++i) {
+    csum_bl.clear();
+    local_obj_absent = false;
+    remote_obj_absent = false;
+
+    ldout(cct, 20) << "obj name: " << ictx->get_object_name(i) << dendl;
+    r = io_ctx.checksum(ictx->get_object_name(i), LIBRADOS_CHECKSUM_TYPE_CRC32C,
+                        init_value_bl, ictx->get_object_size(), 0, 0, &csum_bl);
+    if (r == -ENOENT) {
+      local_obj_absent = true;
```

Comment:
nit: One thought I had: If the local_obj_absent = true, should you continue to checksum the remote? How much faster would a read/other way to see if it exists be? It may be negligible for a single object, but compounding with many absent.

Reply:
I doubt I got you right here, you are asking to check for image existence if the local object is absent instead of performing `remote_io_ctx.checksum()` ?

If I got you right, I'm not sure what would be the benefit as I expect `remote_io_ctx.checksum()` to check for the object existence before anything internally